### ISSUE TITLE
Slow solve functions

### ISSUE THREAD

**Issue by fredrik-johansson** [wbhart/flint2#515]
Several functions only use FFLU, not switching to the Dixon solver for large matrices. This includes:

* fmpz_mat_solve
* fmpz_mat_inv
* fmpq_mat_inv

There is also no fmpq_mat_solve.

Note that fmpq_mat_solve_fmpz_mat does the right thing.

By the way, a related idea: the solve/inv functions for fmpz_mat take a den (or mod, in the case of Dixon) variable as output to express a solution over Q, and return a boolean indicating whether the matrix is invertible over Q.

We could allow passing NULL for den/mod, and in that case, solve over Z and have the boolean indicate whether the matrix is invertible over Z.

**Comment by wbhart** [comment]
I get confused, is NULL different from the zero integer?

**Comment by fredrik-johansson** [comment]
Yes... an fmpz_t argument is passed as a pointer (which could be NULL) to an fmpz (which could be zero) somewhere in memory.

**Comment by wbhart** [comment]
Then this seems like a good idea indeed.

**Comment by fieker** [comment]
If you want to spend time here, I'd actually rename the function to indicate that they are not solving over Z... fmpz_mat_solve_rat or p(seudo)solve

**Comment by wbhart** [comment]
Just to clarify @fredrik-johansson did you mean that you want to solve over Q and if the solution happens to be over Z then return the result, otherwise return fail, or did you mean to do a full solve over Z using HNF and the like?

I'll open a separate ticket for this and close this one once it is clarified.

**Comment by fredrik-johansson** [comment]
The former. Solving with HNF should have a separate function.

**Comment by fieker** [comment]
On Tue, May 12, 2020 at 08:22:38AM -0700, Fredrik Johansson wrote:
> The former. Solving with HNF should have a separate function.
> 
Notabout solve_with_hnf, it's solve as a Z-module problem or solve as a
Q-module problem, The representation is a sperate issue
